components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/engine/CamelSubscription.java

```diff
@@ -0,0 +1,252 @@
+            for (StreamPayload<Exchange> data : sendingQueue) {
+                // TODO what if the subscriber throws an exception?
+                this.subscriber.onNext(data);
+
+                mutex.lock();
+                boolean shouldStop = this.terminated;
+                mutex.unlock();
+
+                if (shouldStop) {
+                    break;
+                }
+            }
+        }
+    }
+
+    public void signalCompletion() throws Exception {
+        mutex.lock();
+        terminating = true;
+        mutex.unlock();
+
+        checkAndFlush();
+    }
+
+    @Override
+    public void cancel() {
+        publisher.unsubscribe(this);
+
+        mutex.lock();
+        this.terminated = true;
+        List<StreamPayload<Exchange>> bufferCopy = new LinkedList<>(buffer);
+        this.buffer.clear();
+        mutex.unlock();
+
+        discardBuffer(bufferCopy);
+    }
+
+    protected void discardBuffer(List<StreamPayload<Exchange>> remaining) {
+        for (StreamPayload<Exchange> data : remaining) {
+            data.getCallback().processed(data.getItem(), new IllegalStateException("Cannot process the exchange " + data.getItem() + ": subscription cancelled"));
+        }
+    }
+
+    public void publish(StreamPayload<Exchange> message) {
+        StreamPayload<Exchange> discardedMessage = null;
+        String discardReason = null;
+        try {
+            mutex.lock();
+            if (!this.terminating && !this.terminated) {
+                if (this.backpressureStrategy == ReactiveStreamsBackpressureStrategy.BUFFER) {
+                    buffer.addLast(message);
+                } else if (this.backpressureStrategy == ReactiveStreamsBackpressureStrategy.DROP) {
+                    if (buffer.size() > 0) {
+                        LOG.warn("Exchange " + message.getItem() + " dropped according to the backpressure strategy " + ReactiveStreamsBackpressureStrategy.DROP);
+                        discardedMessage = message;
+                        discardReason = "the backpressure strategy (DROP) does not allow buffering";
+                    } else {
+                        buffer.addLast(message);
+                    }
+                } else if (this.backpressureStrategy == ReactiveStreamsBackpressureStrategy.LATEST) {
+                    if (buffer.size() > 0) {
+                        StreamPayload<Exchange> older = buffer.removeFirst();
+                        LOG.warn("Exchange " + message.getItem() + " dropped according to the backpressure strategy " + ReactiveStreamsBackpressureStrategy.LATEST);
+                        discardedMessage = older;
+                        discardReason = "the backpressure strategy (LATEST) does not allow buffering";
+                    }
+                    buffer.addLast(message);
+                } else {
+                    throw new IllegalStateException("Unsupported backpressure strategy: " + this.backpressureStrategy);
+                }
```

Comment:
Would be nice if ReactiveStreamsBackpressureStrategy were an Interface so that one can maybe plug a custom strategy.